src/btree/bt_sync_obsolete.c

```
@@ -290,3 +295,372 @@ __wt_sync_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
+
+    /* Loop through the eligible candidates. */
+    WT_ERR(cursor->get_key(cursor, &key));
+    /* Check we are still dealing with keys that have the right prefix. */
+    if (!WT_PREFIX_MATCH(key, WT_URI_FILE_PREFIX)) {
+        ret = WT_NOTFOUND;
+        goto err;
+    }
+
+    /* Save the selected uri. */
+    WT_ERR(__wt_buf_set(session, uri, cursor->key.data, cursor->key.size));
+
+err:
+    WT_TRET(__wt_metadata_cursor_release(session, &cursor));
+    return (ret);
+}
+
+/*
+ * __checkpoint_cleanup_int --
+ *     Internal function to perform checkpoint cleanup of all eligible files.
+ */
+static int
+__checkpoint_cleanup_int(WT_SESSION_IMPL *session)
+{
+    WT_DECL_ITEM(uri);
+    WT_DECL_RET;
+
+    WT_RET(__wt_scr_alloc(session, 1024, &uri));
+    WT_ERR(__wt_buf_set(session, uri, WT_URI_FILE_PREFIX, strlen(WT_URI_FILE_PREFIX) + 1));
+
+    while ((ret = __checkpoint_cleanup_get_uri(session, uri)) == 0) {
+        /* Ignore the metadata file. */
+        if (strcmp(uri->data, WT_METAFILE_URI) == 0)
+            continue;
+
+        ret = __checkpoint_cleanup_walk_btree(session, uri);
+        if (ret == ENOENT || ret == EBUSY) {
+            __wt_verbose_debug1(session, WT_VERB_CHECKPOINT_CLEANUP,
+              "%s: skipped performing checkpoint cleanup because the file %s", (char *)uri->data,
+              ret == ENOENT ? "does not exist" : "is busy");
+            continue;
+        }
+        WT_ERR(ret);
+
+        /*
+         * Wait here for some time before proceeding with another table to minimize the impact of
+         * checkpoint cleanup on the regular workload.
+         */
+        __wt_cond_wait(session, S2C(session)->cc_cleanup.cond,
+          WT_CHECKPOINT_CLEANUP_FILE_INTERVAL * WT_MILLION, __checkpoint_cleanup_run_chk);
+
+        /* Check if we're quitting. */
+        if (!__checkpoint_cleanup_run_chk(session))
+            break;
+    }
+
+err:
+    __wt_scr_free(session, &uri);
+    return (ret == WT_NOTFOUND ? 0 : ret);
```

I don't think we should have `WT_NOTFOUND` here, after the `err` label. If you do this, you may hide errors that we should not have.
I would do this before the `err` label:
```
if (ret == WT_NOTFOUND) ret = 0;
```

---

thanks, fixed.